### PR TITLE
Support modified_beam_search with hotwords for streaming NeMo transducer models

### LINKED ISSUE #3572: Support modified_beam_search (and hotwords) for Nemotron streaming transducer
### Summary

I was benchmarking for a long-form transcription (streaming) project and hit a bit of a dead end... 

`OnlineRecognizer.from_transducer` with `model_type="nemotron"` only supports `decoding_method="greedy_search"`. Passing `modified_beam_search` errors out at:

```
csrc/online-recognizer-transducer-nemo-impl.h:OnlineRecognizerTransducerNeMoImpl:56
Unsupported decoding method: modified_beam_search
```

Because `hotwords_file` and `hotwords_score` are tied to `modified_beam_search`, this sadly leaves the Nemotron streaming transducer with no hotwords / contextual-biasing path. For long-form streaming transcription that includes proper nouns or domain-specific vocabulary (technical terms, place names, person names that aren't in the BPE as units), the model substitutes phonetic neighbours that the hotwords API would otherwise be able to bias correctly.

### Why this matters

`nemotron-speech-streaming-en-0.6b` (PR #3555, 2026-04-25 export) is currently the strongest English streaming transducer in `sherpa-onnx` for low-latency transcription tasks. Tested side-by-side on a 62-second long-form speech sample:

- **Whisper Turbo (offline, beam=5)**: WER 11.6%, RTF 0.140
- **Nemotron jan2026 streaming + tuned endpoints**: WER 12.6%, RTF 0.036

Streaming Nemotron is ~4× faster than offline Whisper-Turbo at an almost-equal WER, with the streaming property that offline Whisper can't provide. The remaining 1-word gap on this sample is a single proper-noun substitution (`re-extract` → `rextract`) that hotwords would have caught. Closing that gap on common domain vocabulary would make the model production-ready for a class of applications (meeting transcription, lecture capture, podcast indexing, dictation) where Whisper is currently chosen primarily for the hotwords path.

### What works today (for reference)

`hotwords_file` works correctly with Zipformer streaming transducers and offline Conformer / Paraformer recognisers when paired with `modified_beam_search`. The C++ implementation `OnlineRecognizerTransducerImpl` (non-Nemo) already supports both decoding methods. The Nemotron variant lives at `online-recognizer-transducer-nemo-impl.h` and explicitly rejects anything other than `greedy_search`.

### Asking for

Extension of `OnlineRecognizerTransducerNeMoImpl` to support `modified_beam_search` (and through it, `hotwords_file` + `hotwords_score`). The Nemotron model's transducer architecture is structurally similar to other streaming transducers `sherpa-onnx` already supports, so the work is plausibly extending the existing modified_beam_search code path rather than a from-scratch implementation.

### Test reproduction (offline, no audio needed)

```python
import sherpa_onnx

# Any of the 2026-04-25 Nemotron variants downloaded from the
# k2-fsa releases page
md = "models/sherpa-onnx-nemotron-speech-streaming-en-0.6b-1120ms-int8-2026-04-25"
sherpa_onnx.OnlineRecognizer.from_transducer(
    tokens=f"{md}/tokens.txt",
    encoder=f"{md}/encoder.int8.onnx",
    decoder=f"{md}/decoder.int8.onnx",
    joiner=f"{md}/joiner.int8.onnx",
    sample_rate=16000,
    feature_dim=128,
    model_type="nemotron",
    decoding_method="modified_beam_search",  # → errors
)
```

Tested on `sherpa-onnx==1.13.0`.

Happy to test any branch or PR with the same long-form speech samples + WER harness if it'd help validate! ^_^

### PR DESCRIPTION
**Pull request by josephomills**
Fixes #3572

## What this PR does

`OnlineRecognizerTransducerNeMoImpl` only supports `greedy_search`, so streaming NeMo/Nemotron models have no hotwords / contextual-biasing path (hotwords are gated on `modified_beam_search`). This PR ports `OfflineTransducerModifiedBeamSearchNeMoDecoder` (#3859-era code) to streaming:

- **`OnlineTransducerModifiedBeamSearchNeMoDecoder`** — the offline decoder's frame-asynchronous beam loop, restructured to run chunk by chunk. Between chunks the active hypotheses are persisted in the stream result's existing `hyps` field; each hypothesis carries its own copy of the stateful prediction-network states in a new `Hypothesis::nemo_decoder_states` field (`CopyableOrtValue`, the same mechanism already used for LM shallow-fusion states).
- **Hotwords** work through the standard context-graph machinery: recognizer-level `hotwords_file` / `hotwords_buf` and per-stream `CreateStream(hotwords)`. The pre-top-k boost is used for candidate ranking only (original scores are restored right after), so path scores and `ys_probs` stay purely acoustic and the context-graph score enters exactly once via `ForwardOneStep()`. (The offline decoder double-counts the boost here — same pattern, happy to fix it there in a follow-up.)
- A small `OnlineTransducerNeMoDecoder` interface is introduced so the impl can hold either decoder; the greedy decoder just gains the base class.

### One deliberate improvement over the offline decoder

During verification on long real audio, the direct port showed a deletion bias: the beam fills up with alignments of one and the same token sequence, and paths that emit tokens in low-confidence regions get pruned in favor of blank-only paths, permanently dropping quiet phrases (WER 35.8% vs greedy's 23.2% on a 2-minute test clip). The fix recombines candidates that share (token sequence, frame, per-frame symbol count) each pruning round, keeping the best-scoring alignment — their decoder and context states are deterministic functions of the emitted tokens. With recombination, beam=4 **beats** greedy (22.1% vs 23.2%) on the same clip. (Recombining with log-sum-exp — i.e. marginalizing over emission positions — was tried first and gives the same WER, but it lets a weak, consistently-probable token accumulate mass across positions and beat silence on noise-only audio, so the maximum is kept instead, matching greedy and the offline decoder.) The offline decoder has the same latent pruning bias; I can apply the same fix there in a follow-up if wanted.

## Measurements

Streaming Nemotron 3.5 1120 ms int8, macOS arm64 CPU (2 threads), exercised through the node addon:

- **WER on 120 s of real sermon audio** (vs a whisper-large pseudo-reference): greedy 23.16%, modified_beam_search beam=4 **22.11%**. Wall cost ~1.1× greedy on that clip, RTF ≈ 0.11 — the decoder output is cached per hypothesis, so the prediction network runs only for tokens whose hypothesis survived pruning, and the encoder dominates.
- **Beam recovered a proper noun greedy mangled** with no hotwords at all: "Jehoshaphat" (greedy: "Jehoo").
- **Hotword biasing** on a clip containing six rare biblical names (score 2.0, `bpe_vocab` derived from tokens.txt as in #3723): fixed "Melchizedek" (greedy: "Melk Izadek"), "Zerubbabel" ("Zarabbable"), "Epaphroditus" ("Epoditus"), and "Jeshua" — exactly the rare-domain-term failure class NVIDIA's word-boosting guidance targets.
- **Streaming invariant**: chunked feeding produces byte-identical output to whole-utterance feeding.
- `ys_probs` and `context_scores` are per-token aligned; `num_trailing_blanks` tracked per hypothesis, so endpointing works unchanged.
- Language prompts for multilingual models are unaffected (applied in the impl's encoder call, decoder-agnostic).

## Changes

- `sherpa-onnx/csrc/online-transducer-modified-beam-search-nemo-decoder.{h,cc}` — the new decoder.
- `sherpa-onnx/csrc/online-transducer-nemo-decoder.h` — common decoder interface for streaming NeMo transducers.
- `sherpa-onnx/csrc/online-transducer-greedy-search-nemo-decoder.h` — derives from the new interface (no behavior change).
- `sherpa-onnx/csrc/hypothesis.h` — new `nemo_decoder_states` field.
- `sherpa-onnx/csrc/online-recognizer-transducer-nemo-impl.h` — `modified_beam_search` branch in both constructors, hotwords/bpe initialization, `CreateStream(hotwords)` (all mirroring `OnlineRecognizerTransducerImpl`).
- `sherpa-onnx/csrc/CMakeLists.txt` — new source file.

Downstream context: FreeShow, an open-source church presentation app with an AI scripture feature under review (ChurchApps/FreeShow#3579), runs streaming Nemotron via sherpa-onnx-node and needs hotword biasing for biblical vocabulary on live transcripts.

## Notes

- The reported hypothesis is selected by raw score, matching the offline decoder — not by length-normalized score: a blank-only hypothesis (legitimate for silence) has an empty token sequence, and length normalization lets token paths amortize their cost and beat silence on noisy audio.
- Silence behavior: after real speech, trailing silence adds no tokens (identical to greedy, verified over 8 s). A stream consisting only of noise from the very start, decoded in auto-language mode, can commit a language tag plus a stray token — the sequence-level likelihood of this model genuinely prefers early tag commitment there (greedy avoids it only by being locally myopic, and the offline decoder shares the property by mechanism). Language pinning or VAD gating avoids it.
- Greedy behavior is untouched (verified same output before/after).
- With beam search, the best path can revise earlier text between chunks (inherent to beam) — greedy's monotonic-prefix property does not hold.
- Nemotron token sets contain no `<unk>`, so no unk handling is included.
- As with the offline decoder, raw-word hotwords need `modeling_unit: bpe` plus a `bpe_vocab` (derivable from tokens.txt, see #3723); pre-tokenized hotwords work without it.
- `./scripts/check_style_cpplint.sh 1` passes. Built with the `test-nodejs-addon-api.yaml` cmake flags. The related CI workflows do not run on pull requests; verification above is local.

## Summary by CodeRabbit

* **New Features**
  * Added streaming modified beam-search decoding for NeMo transducer models.
  * Added configurable hotwords from files, buffers, and individual streams.
  * Added BPE-encoded hotword support with boost scoring.
  * Preserved decoding hypotheses across streaming chunks for improved continuity.
  * Added configurable beam limits and blank-penalty controls.
  * Added BPE vocabulary and modeling-unit options to the file-decoding example.
  * Improved decoding efficiency by reusing prediction results during hypothesis expansion.
* **Tests**
  * Added coverage for NeMo modified beam-search decoding with and without hotwords.

### CODE REVIEW

File: sherpa-onnx/csrc/online-transducer-modified-beam-search-nemo-decoder.cc [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        auto it = index.find(key);
+        if (it == index.end()) {
+          index[key] = merged.size();
+          merged.push_back(std::move(p));
+        } else {
+          auto &dst = merged[it->second];
+          dst.second.hyp.log_prob = LogAdd<double>()(dst.second.hyp.log_prob,
+                                                     p.second.hyp.log_prob);
+          dst.first = dst.second.hyp.log_prob;
+        }
```
[the remark this answers is not in the document]
Comment (josephomills) [reply sherpa-onnx/csrc/online-transducer-modified-beam-search-nemo-decoder.cc]: Addressed in f6c903d5 — the recombination identity now includes num_symbols, and the merge keeps the best-scoring alignment outright (Viterbi-style) rather than log-sum-exp, so the surviving metadata (timestamps etc.) is the representative path's by construction and does not depend on iteration order. With the in-loop dedup, candidates reaching the chunk boundary all have distinct token sequences (frame == num_rows, num_symbols == 0 for all), so hyps.Add() no longer merges anything there. The switch from summing to max also matters behaviorally: marginalizing over emission positions let a weak but consistently-probable token accumulate mass and beat silence on noise-only audio; keeping the maximum matches greedy and the offline decoder.

File: sherpa-onnx/csrc/online-transducer-modified-beam-search-nemo-decoder.cc [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+  // Persist the hypotheses for the next chunk. Hypotheses with an identical
+  // token sequence are merged; their decoder states are identical since the
+  // prediction network is a deterministic function of the emitted tokens.
+  Hypotheses hyps;
+  for (auto &c : cur) {
+    hyps.Add(std::move(c.hyp));
+  }
+
+  Hypothesis best = hyps.GetMostProbable(true /*length_norm*/);
```
[the remark this answers is not in the document]
Comment (josephomills) [reply sherpa-onnx/csrc/online-transducer-modified-beam-search-nemo-decoder.cc]: Addressed in f6c903d5 by taking your second suggestion: the best hypothesis is selected by raw score (GetMostProbable(false)), matching the offline decoder's max_element selection, so the empty blank-only hypothesis is never divided by zero. Length normalization turned out to be independently harmful here — a token path amortizes its cost over its length while the blank-only path cannot, letting weak tokens beat legitimate silence on noisy audio. Verified: 8 s of silence after real speech adds no tokens (identical to greedy), and WER on the 2-minute long-form clip is unchanged (22.1% vs greedy 23.2%). The one remaining edge, noted in the PR description, is a stream that is noise-only from the very start in auto-language mode, where the model's sequence-level likelihood genuinely prefers committing a language tag early; language pinning or VAD gating avoids it.

**Comment by josephomills** [comment]
@coderabbitai Both outside-diff comments are addressed in 97f3825f:

1. **Decoder output lifecycle** — adopted: the decoder output is now cached per hypothesis (`Hypothesis::nemo_decoder_out`). Blank transitions and later frames reuse it with no decoder call, and a newly emitted token is consumed lazily at its next expansion, only if its candidate survived pruning. This cut the 2-minute long-form decode by ~18% (beam=4 is now ~1.1× greedy). One correction to the finding's premise, for the record: the previous code did not consume tokens twice — it kept the states one token behind and fed `ys.back()` exactly once at the next expansion, the same lifecycle as the offline beam decoder, which the byte-identical chunked-vs-whole results reflected. The rework is a performance improvement and an alignment with the greedy decoder's lifecycle, not a correctness fix.

2. **Hotword boost separation** — confirmed and fixed: the boost is applied to the logits only for the top-k ranking and the original values are restored immediately after, so `log_prob` and `ys_probs` stay purely acoustic and the context score enters exactly once via `ForwardOneStep()`. Note the offline NeMo beam decoder has the same double-count (this pattern was ported from there) — happy to fix it there in a follow-up PR.

Re-verified after both changes: WER 22.11% (unchanged; greedy 23.16%), chunked == whole byte-identical, hotword name corrections intact, post-speech silence identical to greedy, `./scripts/check_style_cpplint.sh` passes.

**Comment by josephomills** [comment]
Hi @csukuangfj, following up on this PR. I'm integrating it into FreeShow, an open-source presentation app, so I'd be happy if you could help 🙏

This PR gives streaming NeMo models a contextual-biasing path, which they have no way to get today, and closes #3572. Boosting repaired rare proper nouns that greedy breaks apart, and it performs better on WER.

This may also be relevant to the silence fixes you merged this week (#3920, #3921, #3923, #3924): after real speech, trailing silence adds no tokens with this decoder, the same as greedy.

**Review by csukuangfj — COMMENTED** [review]
Can you add CI tests to cover your changes and update copyright information in the header of each file?

**Comment by josephomills** [comment]
@csukuangfj Done in ad351004.

CI tests: added modified_beam_search runs, with and without hotwords, to `.github/scripts/test-online-transducer.sh`, run under both `sherpa-onnx` and `decode-file-c-api` (which gained `--modeling-unit` and `--bpe-vocab` for the hotwords run).

Copyright: updated the header of the three new files and added my line to `online-recognizer-transducer-nemo-impl.h`; the remaining three files only have one-line to nine-line edits, so I left their headers unchanged.

Also merged master into the branch, and the recognizer now exits with a clear message if `--lm` is given with this decoder, since LM fusion is not implemented for it (it was silently ignored before).

File: sherpa-onnx/csrc/online-transducer-modified-beam-search-nemo-decoder.cc [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+      // The boost is used for ranking only: the original values are restored
+      // right after, so path scores and ys_probs stay purely acoustic and
+      // the context-graph score is added exactly once via ForwardOneStep().
+      std::vector<std::pair<int32_t, float>> boosted;
+      if (context_graph != nullptr && c.hyp.context_state != nullptr) {
+        for (const auto &pair : c.hyp.context_state->next) {
+          int32_t token_id = pair.first;
+          if (token_id >= 0 && token_id < vocab_size) {
+            boosted.emplace_back(token_id, p_logit[token_id]);
+            p_logit[token_id] += hotwords_score;
```
[the remark this answers is not in the document]
Comment (josephomills) [reply sherpa-onnx/csrc/online-transducer-modified-beam-search-nemo-decoder.cc]: Fixed in 186c1157: the ranking boost now uses the transition's own token_score (read from the child node; the same value ForwardOneStep returns for a direct match), so per-hotword scores rank consistently with path scoring. The now-unused hotwords_score parameter was removed from the decoder.

**Comment by josephomills** [comment]
@coderabbitai The hotwords test now asserts on the result in 186c1157: the sherpa-onnx run must show the hotwords score in context_scores for the matched tokens (the C API tool prints only the text, so the check is skipped there). A transcript comparison would not catch a regression here, since the hotwords are words the model already recognizes.

File: sherpa-onnx/csrc/online-transducer-modified-beam-search-nemo-decoder.cc [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        all_candidates.emplace_back(log_prob, std::move(c));
+        continue;
+      }
+
+      if (c.num_symbols >= max_symbols_per_frame) {
+        // Reached the per-frame symbol limit; force advancing to the next
+        // frame without emitting a token
+        c.frame += 1;
+        c.num_symbols = 0;
+        c.hyp.num_trailing_blanks += 1;
```
[the remark this answers is not in the document]
Comment (josephomills) [reply sherpa-onnx/csrc/online-transducer-modified-beam-search-nemo-decoder.cc]: Fixed in ebf10dca. The counter is now left unchanged on the forced advance, matching the greedy decoder.

Measured with --blank-penalty=20, which saturates the cap on every frame: greedy reports 0 trailing blanks, this decoder reported 1 before the change and 0 after. Normal decoding is byte-identical before and after.

**Review by csukuangfj — APPROVED** [review]
Thank you for your contribution!
